### PR TITLE
Changes to Gradient Mixin and Fitness Function Tests for Pybind Objects

### PR DESCRIPTION
**Pull request by nightdr**
I changed the gradient mixin and fitness function tests to work with pybind/c++ objects.

### CODE REVIEW

**Comment by nightdr** [comment]
Also added copy tests for implicit and explicit regression to test for segmentation faults with pybind objects.

File: tests/unit/evaluation/test_fitness_function.py [hunk 1]
```diff
@@ -3,56 +3,124 @@
...
-    mocker.patch.object(VectorBasedFunction, "evaluate_fitness_vector",
+def test_vector_based_function_invalid_metric(engine, mocker, vector_based_function):
+    if engine == "Python":
+        mocker.patch.object(vector_based_function, "__abstractmethods__",
+                            new_callable=set)
+    mocker.patch.object(vector_based_function, "evaluate_fitness_vector",
                         return_value=[-2, -1, 0, 1, 2])
-    with pytest.raises(KeyError):
-        _ = VectorBasedFunction(metric="invalid metric")
+    with pytest.raises((KeyError, ValueError)):  # KeyError in Python, ValueError in Cpp
```
Comment (gbomarito) [inline tests/unit/evaluation/test_fitness_function.py:110]: Can we change the python error type to be consistent with c++?

File: tests/unit/evaluation/test_gradient_mixin.py [hunk 2]
```diff
@@ -72,20 +95,28 @@ class VectorGradientMixinWithNewParent(vector_gradient_mixin, NewParent):
         def get_fitness_vector_and_jacobian(self, individual):
             pass
 
-    with pytest.raises(KeyError):
+    with pytest.raises((KeyError, ValueError)):  # KeyError in Python, ValueError in Cpp
```
Comment (gbomarito) [inline tests/unit/evaluation/test_gradient_mixin.py:98]: Same as above comment

File: tests/unit/symbolic_regression/test_explicit_regression.py [hunk 3]
```diff
@@ -310,3 +312,10 @@ def test_explicit_regression_relative_get_fit_vec_and_jac(
 
 def test_can_pickle(sample_regression):
     _ = dill.loads(dill.dumps(sample_regression))
+
+
+def test_can_copy(sample_regression):
```
Comment (gbomarito) [inline tests/unit/symbolic_regression/test_explicit_regression.py:317]: does this test automatically test the desructors you implemented?

**Review by gbomarito — CHANGES_REQUESTED** [review]
Looks good, just a minor update requested

File: tests/unit/evaluation/test_fitness_function.py (hunk 1 above)
Comment (nightdr) [reply tests/unit/evaluation/test_fitness_function.py:110]: It's relatively easy to make the Python error match the C++ error and vice versa. Let me know which way you prefer (match Python or match C++), and I can fix that asap

File: tests/unit/symbolic_regression/test_explicit_regression.py (hunk 3 above)
Comment (nightdr) [reply tests/unit/symbolic_regression/test_explicit_regression.py:317]: It tests the issue that was causing the segmentation fault before, which was solved by changing the destructors. But this test doesn't test for any memory leaks/that the training data was deleted correctly if that answers your question.

File: tests/unit/evaluation/test_fitness_function.py (hunk 1 above)
Comment (gbomarito) [reply tests/unit/evaluation/test_fitness_function.py:106]: lets go with the ValueError

File: tests/unit/symbolic_regression/test_explicit_regression.py (hunk 3 above)
Comment (gbomarito) [reply tests/unit/symbolic_regression/test_explicit_regression.py:317]: yep, thats clear now, thanks

File: tests/unit/evaluation/test_fitness_function.py (hunk 1 above)
Comment (nightdr) [reply tests/unit/evaluation/test_fitness_function.py]: sounds good, just changed that

File: tests/unit/symbolic_regression/test_explicit_regression.py (hunk 3 above)
Comment (nightdr) [reply tests/unit/symbolic_regression/test_explicit_regression.py:317]: of course no problem

**Pull request merged**